### PR TITLE
erc20 impl with first test

### PR DESCRIPTION
**Pull request by gwalen**
The commits contain:
- erc20 abstract impl
- concrete implementation of Erc20 in form of MyToken 
- examples to check if MyToken works fine
- early tests ( this is WIP) 

---> update
test are fixed, included also test for error cases

### CODE REVIEW

File: .DS_Store
Comment (alexfertel) [inline .DS_Store:1]: This file should not be included in version control.

File: examples/my_token.rs [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+        erc20_token_address: my_token_address,
+    })
+}
+
+fn read_secret_from_file(fpath: &str) -> eyre::Result<String> {
+    let f = std::fs::File::open(fpath)?;
+    let mut buf_reader = BufReader::new(f);
+    let mut secret = String::new();
+    buf_reader.read_line(&mut secret)?;
+    Ok(secret.trim().to_string())
```
Comment (alexfertel) [inline examples/my_token.rs]:
```suggestion
    std::fs::read_to_string(fpath)
```

File: examples/my_token.rs [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+    let alice_key_path = std::env::var(ALICE_PRIV_KEY_PATH)
+        .map_err(|_| eyre!("No {} env var set", ALICE_PRIV_KEY_PATH))?;
+    let rpc_url = std::env::var(RPC_URL).map_err(|_| eyre!("No {} env var set", RPC_URL))?;
+    let bob_key_path = std::env::var(BOB_PRIV_KEY_PATH)
+        .map_err(|_| eyre!("No {} env var set", BOB_PRIV_KEY_PATH))?;
+
+    let provider = Provider::<Http>::try_from(rpc_url)?;
+    let my_token_address: Address = program_address.parse()?;
+
+    let alice_private_key = read_secret_from_file(&alice_key_path)?;
```
Comment (alexfertel) [inline examples/my_token.rs:208]: Why not keep this in the `.env`?

File: examples/my_token.rs [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+    let mint_tx: TransactionReceipt = my_token
+        .mint(to, 1000.into())
+        .send()
+        .await?
+        .await?
+        .expect("Mint tx returned non");
+    println!("mint tx: {:?}", mint_tx.transaction_hash);
+
+    let deployer_balance: U256 = my_token.balance_of(to).call().await?;
+    println!("mint 'to' balance after : {}", deployer_balance);
```
Comment (alexfertel) [inline examples/my_token.rs:109]: Should these be assertions?

File: tests/erc20_base.rs [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+        .mint(to, amount)
+        .send()
+        .await
+        .unwrap()
+        .await
+        .unwrap()
+        .expect("Mint tx returned non");
+
+    Ok(mint_tx)
+}
```
Comment (alexfertel) [inline tests/erc20_base.rs]: Why do we use a `Result` here if we're `unwrapping`/`expecting`?

File: tests/erc20_base.rs [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+async fn mint(erc20_token: &MyTokenType, to: Address, amount: U256) -> eyre::Result<TransactionReceipt> {
+    let mint_tx: TransactionReceipt = erc20_token
+        .mint(to, amount)
+        .send()
+        .await
+        .unwrap()
+        .await
+        .unwrap()
+        .expect("Mint tx returned non");
```
Comment (alexfertel) [inline tests/erc20_base.rs]: Typo?

Comment (alexfertel) [reply tests/erc20_base.rs]: Also, why do we `expect` after an `unwrap`, is this a `Result<Result<...>>`?

File: tests/erc20_base.rs [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+    bob_client: SignerClient,
+    erc20_token_address: Address,
+}
+
+static FIXTURES: OnceCell<Mutex<Fixtures>> = OnceCell::const_new();
+
+#[tokio::test]
+async fn mint_test() {
+    let fixtures_mutex = init_fixtures().await.unwrap();
+    let fixtures = fixtures_mutex.lock().await;
```
Comment (alexfertel) [inline tests/erc20_base.rs:90]: Really like that you used a mutex here.

File: tests/erc20_base.rs
Comment (alexfertel) [inline tests/erc20_base.rs:1]: I think `examples/my_token.rs` and `tests/erc20_base.rs` are almost the same file. Do we need to keep both?

**Review by alexfertel — COMMENTED** [review]
I know this is mostly setup, so I'm only asking stuff to improve my understanding of the code.

File: examples/my_token.rs (hunk 1 above)
Comment (gwalen) [reply examples/my_token.rs]: much nicer thanks :)

File: examples/my_token.rs (hunk 2 above)
Comment (gwalen) [reply examples/my_token.rs:208]: For the sake of this project we definitely can put that in `.env`
In other projects I was keeping the keys in the separate location to to accidentally copy & paste them when sharing contents of `.env` with other developers.

File: examples/my_token.rs (hunk 3 above)
Comment (gwalen) [reply examples/my_token.rs:109]: contents of  examples/my_example.rs are just for my internal quick prototyping and testing, I will remove them 
after the implementation is finished, that is why this file is a little mess I should have said that earlier :)

File: tests/erc20_base.rs
Comment (gwalen) [reply tests/erc20_base.rs:1]: contents of examples/my_example.rs are just for my internal quick prototyping and testing, I will remove them
after the implementation is finished, you are right in final structure `examples` should be removed 💯

File: examples/my_token.rs (hunk 2 above)
Comment (alexfertel) [reply examples/my_token.rs:208]: Makes sense 👍🏻

File: examples/my_token.rs (hunk 3 above)
Comment (alexfertel) [reply examples/my_token.rs:109]: no worries hehe

File: tests/erc20_base.rs (hunk 4 above)
Comment (gwalen) [reply tests/erc20_base.rs]: I fixed in last commit, converting Option<T> directly to Result<T>

File: tests/erc20_base.rs (hunk 5 above)
Comment (gwalen) [reply tests/erc20_base.rs]: I fixed in last commit, converting Option directly to Result

**Review by alexfertel — APPROVED** [review]
Looking very good! 💯

**Pull request merged**